### PR TITLE
29344 cython stub generator

### PR DESCRIPTION
**Pull request by teodoramilanovic**
Hey @sshane, @jnewb1, @adeebshihadeh,  

[CythonPEG](https://github.com/RaubCamaioni/CythonPEG) is a parser that was specifically developed for the automatic generation of Python stub files from Cython files. I forked it and implemented several changes and additions to enhance its usability and functionality. Below is the summary of its functionalities including both the original features and the ones that have been added or modified:
- Object definition,
- Expression definition,
- Class definitions,
- Function definitions,
- Struct definition,
- Enum definitions, 
- Alias definition,
- Import statement definition, 
- External declaration definition, 
- Compiler directives definition, 
- Docstring definition,
- Recursive definitions. 

I've also implemented a script that utilizes this parser to generate stubs (.pyi) from Cython files (.pyx, .pxd). It has the capability to parse either all relevant Cython files within the openpilot repository or specific ones provided as arguments via the command line. I've also utilized the script to generate stubs and place each of them in a 'stubs' directory alongisde the original file. I'm looking forward to your feedback to ensure that this is something you wanted for this issue. Please let me know if there are any changes you would like me to make.

Thanks!

### CODE REVIEW

**Comment by adeebshihadeh** [comment]
Diff is unreadable here. Can you gitignore the generated code so I can read the diff?

**Comment by teodoramilanovic** [comment]
> Diff is unreadable here. Can you gitignore the generated code so I can read the diff?

Hey, I ignored .pyi files, as you requested. If you want to see how they look I can post them here.

File: scripts/parse_cython_to_stub.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+def read_parse_and_write(file_path):
+    with open(file_path, 'r') as file:
+        cython_tokens = file.read()
+
+    stub_file, unparsed_tokens = cp.cython_string_2_stub(cython_tokens)
+    file_path = os.path.splitext(file_path)[0]
+    last_slash_index = file_path.rfind('/')
+    new_path = file_path[:last_slash_index]
+    stubs_path = new_path + '/stubs'
```
[the remark this answers is not in the document]
Comment (teodoramilanovic) [reply scripts/parse_cython_to_stub.py:24]: I wasn't exactly sure where to put the generated stub files. I'd be grateful if you could advise me on the appropriate location for them. Also, 'libcpp' is included in 'cimport' statements, I can ignore those while parsing Cython files if that works for you?

[the remark this answers is not in the document]
Comment (sshane) [reply scripts/parse_cython_to_stub.py:24]: Sounds like they should be in the same directory as the source files

**Comment by adeebshihadeh** [comment]
Indent is still wrong. This PR has required more review than it appears to be worth.

Feel free to open a new one once it's cleaned up and ready to go.

**Pull request closed without merging**